### ISSUE TITLE
Validation Tests Error

### ISSUE THREAD

**Issue by andersonfrailey** [open-source-economics/Tax-Calculator#1231]
I'm working on a new branch and when I ran the validation test this traceback error popped up:
```
STARTING WITH VALIDATION TESTS : Thu Mar  9 16:00:52 EST 2017
Traceback (most recent call last):
  File "simtax.py", line 104, in <module>
    sys.exit(main())
  File "simtax.py", line 96, in main
    output_records=args.records)
  File "/Users/andersonfrailey/Tax-Calculator/taxcalc/validation/taxsim/simpletaxio.py", line 117, in __init__
    output_records)
  File "/Users/andersonfrailey/Tax-Calculator/taxcalc/validation/taxsim/simpletaxio.py", line 609, in _calc_object
    start_year=self._policy.start_year)
TypeError: __init__() got an unexpected keyword argument 'gfactors'
ERROR: first-output-file d15.in.out-simtax does not exist
. taxsim/d15
============ ALL EXECUTED TAXSIM VALIDATION TESTS PASS
FINISHED WITH VALIDATION TESTS : Thu Mar  9 16:01:10 EST 2017
```

As you can see, it says the tests passed, but I'm worried about the error that comes up. Also, after I ran the tests, the content of the file `d15.taxdiffs` was erased. Before the tests I checked to make sure the branch was up-to-date.

When I switch over to my master branch, which is also up-to-date, I get a different error and the tests failed. Here is that error:

```
STARTING WITH VALIDATION TESTS : Thu Mar  9 16:15:20 EST 2017
Traceback (most recent call last):
  File "simtax.py", line 104, in <module>
    sys.exit(main())
  File "simtax.py", line 96, in main
    output_records=args.records)
  File "/Users/andersonfrailey/Tax-Calculator/taxcalc/validation/taxsim/simpletaxio.py", line 117, in __init__
    output_records)
  File "/Users/andersonfrailey/Tax-Calculator/taxcalc/validation/taxsim/simpletaxio.py", line 609, in _calc_object
    start_year=self._policy.start_year)
TypeError: __init__() got an unexpected keyword argument 'gfactors'
ERROR: first-output-file d15.in.out-simtax does not exist
F	taxcalc/validation/taxsim/d15.taxdiffs
TEST FAILURE: any red lines starting with F between the
              STARTING... and FINISHED... lines above
              indicate a test failure caused by the actual
              *.taxdiffs results being different from the
              expected *.taxdiffs results.
============ SOME EXECUTED TAXSIM VALIDATION TESTS FAIL
FINISHED WITH VALIDATION TESTS : Thu Mar  9 16:15:38 EST 2017
``` 

Like on the other branch, the contents of `d1.taxdiffs` was deleted after running the tests.

Has anybody else run into this error?

cc: @martinholmer @MattHJensen

**Comment by martinholmer** [comment]
@andersonfrailey asked in issue #1231:
> Has anybody else run into this error?

No, I have not run into this error.  In fact, when I do the following everything is fine.
```
iMac2:tax-calculator mrh$ cd taxcalc/validation/
iMac2:validation mrh$ ./tests all
STARTING WITH VALIDATION TESTS : Thu Mar  9 16:47:54 EST 2017
. taxsim/a15
. taxsim/d15
============ ALL EXECUTED TAXSIM VALIDATION TESTS PASS
FINISHED WITH VALIDATION TESTS : Thu Mar  9 16:49:06 EST 2017
iMac2:validation mrh$ 
```
The error you are getting is the following:
```
File ".../taxcalc/validation/taxsim/simpletaxio.py", line 609, in _calc_object
    start_year=self._policy.start_year)
TypeError: __init__() got an unexpected keyword argument 'gfactors'
```
But the Records class `__init__()` clearly has a `gfactors` argument.

You had problems before with you local git repo not being in sync with the GitHub Tax-Calculator repo.  My guess is that figuring out how that happen before will give you the answer to your current problem.  Didn't you and Matt talk about this issue yesterday?

**Comment by andersonfrailey** [comment]
@martinholmer I was able to fix the issues I was having yesterday and reset my local project to exactly match the upstream repo. In my copy, the Records class also has a `gfactors` argument and when I run `git status` I get this message:
```
On branch master
Your branch is up-to-date with 'origin/master'.
```
However, the validation tests still fail with the same error message.

**Comment by martinholmer** [comment]
@andersonfrailey said in issue #1231:
> I was able to fix the issues I was having yesterday and reset my local project to exactly match the upstream repo.  In my copy, the Records class also has a gfactors argument and when I run git status I get this message:
>```
>On branch master
> Your branch is up-to-date with 'origin/master'.
>```
> However, the validation tests still fail with the same error message.

I have no idea what your problem is.  If you are really on the master branch, then it is extremely worrying that the code doesn't work the same on your computer and on my computer.  You're working on a Mac, right?  What version of Python are you running, 2.7, 3.4 or 3.5?

I think you need somebody with more git skills than I have to help you.

**Comment by martinholmer** [comment]
@andersonfrailey,  Once PR #1233 is merged into master, sync your local repo with the one at GitHub, and try the validation tests again.  Let me know what happens.

**Comment by martinholmer** [comment]
@andersonfrailey, I think your #1231 problem is caused by your local repo not being in sync with the main GitHub repo.  Here is the reason I think that:

You said that running `git status` on your computer produces this:
```
On branch master
Your branch is up-to-date with 'origin/master'
```
All that means is that your local repo is in sync with **your** GitHub repo.

But your GitHub repo is not in sync with the Open-Source-Economics GitHub repo.  You can see that this is true by going to [your GitHub repo](https://github.com/andersonfrailey/Tax-Calculator) and seeing that it says the following when on the master branch:
```
This branch is 12 commits behind open-source-economics:master.
```

I have no idea how things got to be this way or how to correct the problem.

You need to talk with someone who is more knowledgeable about git and GitHub.

**Comment by MattHJensen** [comment]
@andersonfrailey, I believe this sequence will resolve your problems:

Make sure that `upstream` is pointing at https://github.com/open-source-economics/tax-calculator.git:

`git remote -v`

Checkout the local branch you care about:

`git checkout master` or `git checkout [feature-branch]`

Fetch latest from https://github.com/open-source-economics/tax-calculator.git:

`git fetch upstream`

Merge upstream/master into your local branch that you care about:

`git merge upstream/master`

**Comment by andersonfrailey** [comment]
@martinholmer @MattHJensen thank you both for your help.

I've gotten my GitHub repo in sync and I ran the commands @MattHJensen suggested. The validation tests now pass, but I still get the traceback error. At this point would it just be easiest to wipe the slate clean and create a new fork? None of the solutions I've tried seem to be working and it appears this issue is just local to my machine.

**Comment by MattHJensen** [comment]
> At this point would it just be easiest to wipe the slate clean and create a new fork? None of the solutions I've tried seem to be working and it appears this issue is just local to my machine.

That seems fine as long as it doesn't delete any of your work. 

Before deleting the directory and recreating the directory, you might try to reset your repo to look like upstream/master. This will also delete any work that you've done ahead of upstream/master. 

`git reset --hard upstream/master`

**Comment by andersonfrailey** [comment]
I believe I discovered the root cause of the problem. I unknowingly had a previous version of TaxCalc installed and when I updated all tests passed without issue.

Thanks again for the help!

**Comment by MattHJensen** [comment]
@andersonfrailey, are you using the taxcalc-dev environment when you are doing development work? taxcalc should not be installed in that environment. If it is installed in that environment, you could have conflicts between the source code and the installation.

**Comment by martinholmer** [comment]
@MattHJensen said (paraphrasing):
> When using the `taxcalc-dev` environment to do development work, the `taxcalc` package should **not** be installed on your computer.  If it is installed on your computer, you could have conflicts between the source code and the `taxcalc` package installation.

First, is this correct?  Certainly it squares with my recent experience working with the `tc.py` script in directories outside the `tax-calculator` directory tree.

Second, if correct, shouldn't this be included (even stressed) in the Contributor Guide?

**Comment by andersonfrailey** [comment]
@MattHJensen, I believe I inadvertently did install taxcalc in the taxcalc-dev environment. I've uninstalled so that shouldn't be an issue in the future.

**Comment by MattHJensen** [comment]
@martinholmer, 

There is no problem if the `taxcalc` package is installed on your computer when doing development work in the `taxcalc-dev` environment so long as `taxcalc` is not installed in the `taxcalc-dev` environment itself. 

More generally, the rule is that `taxcalc` should not be installed in the same environment where someone is doing development work. 

Definitely agree that this should be discussed in the contributor guide. I will add a commit to #1230.
